src/HotChocolate/AspNetCore/src/AspNetCore/Serialization/DefaultHttpRequestParser.cs

```diff
@@ -47,6 +51,88 @@ internal class DefaultHttpRequestParser : IHttpRequestParser
             CancellationToken cancellationToken) =>
             ReadAsync(stream, false, cancellationToken);
 
+        public async ValueTask<IReadOnlyList<GraphQLRequest>> ReadMultipartRequestAsync(
+            HttpRequest httpRequest,
+            CancellationToken cancellationToken)
+        {
+            // todo: The IFormCollection is convenient, but it requires us to work with strings instead of a stream
+            var formFeature = new FormFeature(httpRequest);
+            var form = await formFeature.ReadFormAsync(cancellationToken);
+
+            if (form.Count < 2)
+                throw new Exception("At least a 'operations' and 'map' field need to be present.");
+
+            if (!form.TryGetValue("operations", out StringValues operationsCollection) || !operationsCollection.TryPeek(out var operationsString))
+                throw new Exception("No 'operations' specified.");
+
+            if (!form.TryGetValue("map", out StringValues mapCollection) || !mapCollection.TryPeek(out var mapString))
+                throw new Exception("No 'map' specified.");
+
+            var map = JsonSerializer.Deserialize<IDictionary<string, string[]>>(mapString);
+
+            if (map == null)
+                throw new Exception("'map' is not a dictionary of string[].");
+
+            // todo: Stuff like this is just a temporary hack until we have a proper multipart reader
+            var operationsBytes = Encoding.UTF8.GetBytes(operationsString);
+            var operationsStream = new MemoryStream(operationsBytes);
+
+            var requests = await ReadAsync(operationsStream, false, cancellationToken);
+
+            if (requests == null)
+                throw new Exception("Requests could not be parsed.");
+
+            foreach (var request in requests)
+            {
+                // todo: This is quite hacky and requires Utf8GraphQLRequestParser to actually return this
+                //       I choose to do it this way as to not change the existing APIs.
+                //       For a final implementation this should be done differently
+                if (!(request.Variables is Dictionary<string, object?> mutableVariables))
+                    continue;
+
+                foreach (var (key, values) in map)
```

Comment:
This only considers top-level variables. 

What if the Upload scalar is used in an input opject?

```graphql
input FooInput {
   bar: Upload
}
```

Reply:
> Nesting files **anywhere** within operations (**usually** in variables).

Reply:
To be honest, I don't think I will be able to implement usage inside an 'inline' input object directly. My current knowledge of the concepts in this project and the code overall is not enough for this.

I could probably handle the scalar being nested inside a variable using Reflection, but I'm not even sure whether this would be going in the right direction or not.

Reply:
I can do this no worry.